### PR TITLE
feat: Add Debian 11 machine image

### LINKED ISSUE #194: Add support for a Debian 11 machine image
### Code of Conduct

- [X] I have read and agree to the project's Code of Conduct.

### Description

Hello,

here i'm proposing the addition of a templated for Debian/Linux.

I'm already in an advanced state of creation of the configuration for the latest LTS: Debian 11

I would kindly ask your support for advancing in the configuration as i'm encountering a strange issue with the keyboard input at installation startup.

I've produced myself a complete and working boot_command that i am already using with packer+vsphere with the official packer examples, but when i plug what i developed on your work and spin up the server, after ISO boot, the boot_command seems to not be input correctly (maybe due to some keyboard misconfiguration) 

Do you have any advice that could help me debugging it futherly?

### Use Case(s)

-

### Potential Configuration

-

### References

-

### PR DESCRIPTION
**Pull request by evilaliv3**
> **Note**
>
> PR template updated by @tenthirtyam.

## Summary of Pull Request

With this pull request, I would like to propose the addition of a Debian 11 machine. The code produced starts from the scripts already loaded in the repository for Ubuntu and preserves the same exact configuration.

## Type of Pull Request

- [ ] This is a bugfix.
- [x] This is an enhancement or feature.
- [ ] This is a code style / formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is something else.
      Please describe:

Closes #194

## Test and Documentation Coverage

- [x] Tests have been completed (for bugfixes / features). (by @tenthirtyam)
- [x] Documentation has been added / updated (for bug fixes / features).

## Breaking Changes?

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

> **Note**
>
> Added by @tenthirtyam.
>
> Current Gaps:
> - Boot from CD-based kickstart vs HTTPS when the data source is toggled to `disk`

### CODE REVIEW

**Comment by evilaliv3** [comment]
@tenthirtyam : i've improved my pull request aligning all the configurations and files where Debian should be listed.

I've not manually edited the build.sh script manually but instead edited the build.yml template so that you will be able just to re-generate the build.s script when you will eventually integrate the pull request.

Thank you

File: builds/linux/debian/11/linux-debian.pkr.hcl [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  vm_version           = var.common_vm_version
+  remove_cdrom         = var.common_remove_cdrom
+  tools_upgrade_policy = var.common_tools_upgrade_policy
+  notes                = "Version: v${local.build_version}\nBuilt on: ${local.build_date}\n${local.build_by}"
+
+  // Removable Media Settings
+  iso_paths    = ["[${var.common_iso_datastore}] ${var.iso_path}/${var.iso_file}"]
+  iso_checksum = "${var.iso_checksum_type}:${var.iso_checksum_value}"
+  http_content = var.common_data_source == "http" ? local.data_source_content : null
+  cd_content   = var.common_data_source == "disk" ? local.data_source_content : null
```
Comment (tenthirtyam) [inline builds/linux/debian/11/linux-debian.pkr.hcl:84]: Have tests been run using both `http_content` and `cd_content` methods by changing `http` to `disk` in the common vars?

**Review by tenthirtyam — COMMENTED** [review]
No immediate concerns after reviewing the changes - follows the project de facto standards. I'll review the Ansible portions soon and aim to skip this into the next release.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

File: builds/linux/debian/11/linux-debian.pkr.hcl (hunk 1 above)
Comment (evilaliv3) [reply builds/linux/debian/11/linux-debian.pkr.hcl:84]: No actually, i left this as it was.

Can you explain me in more details the effect of this variables and the configurations that you would like me to test?

Comment (tenthirtyam) [reply builds/linux/debian/11/linux-debian.pkr.hcl:84]: I notice that the `data_source_command` is missing in the `locals` and that it's not included in the manifest.

Example:
```hcl
locals {
  #....
  data_source_command = var.common_data_source == "http" ? "url=http://{{ .HTTPIP }}:{{ .HTTPPort }}/ks.cfg" : "file=cdrom:/ks.cfg"
}
```

So that it can be called based on setting `var.common_data_source` to `http` (default) or `disk`.

Example:

```hcl
  boot_command = [
    "<wait3s>c<wait3s>",
    "linux /install.amd/vmlinuz",
    " auto-install/enable=true",
    " debconf/priority=critical",
    " ${local.data_source_command}",
    " noprompt --<enter>",
    "initrd /install.amd/initrd.gz<enter>",
    "boot<enter>"
  ]
```

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

Comment (evilaliv3) [reply builds/linux/debian/11/linux-debian.pkr.hcl:84]: I see, would you like to patch this yourselves?

Comment (evilaliv3) [reply builds/linux/debian/11/linux-debian.pkr.hcl:84]: p.s. i just noticed that all the Ubuntu packer templates misses to use "var.build_password"; Due to this reason the only way to set a password is currently to define it in a configuration file but if the users use the set-var.sh script the variable wont be used.

To replicate it just run a build removing the build_password from the common_configuration file. You will notice that doing that way packer will miss the variable to authenticate even if this will be set with the set-var.sh script.

Comment (tenthirtyam) [reply builds/linux/debian/11/linux-debian.pkr.hcl:84]: I've just now tested the above example whilst I had a moment and the results were as seen below:

image

It seems to find the `ks.cfg` but is hitting an issue. 

Thd project aims to ensure both the `http` and `disk` sources can be used where there may be environments that can't route back to the Packer HTTP server. If you could take a look into it more, I'd appreciate it. Meanwhile, I was looking at the inclusion of the Ansible portions when time is available. I had most of that working last evening after making a few minor edits to your branch.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

Comment (tenthirtyam) [reply builds/linux/debian/11/linux-debian.pkr.hcl:84]: > p.s. i just noticed that all the Ubuntu packer templates misses to use "var.build_password"; Due to this reason the only way to set a password is currently to define it in a configuration file but if the users use the set-var.sh script the variable wont be used.
> 
> To replicate it just run a build removing the build_password from the common_configuration file. You will notice that doing that way packer will miss the variable to authenticate even if this will be set with the set-var.sh script.

Please reference this in a new issue (`bug`). 🐞 

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

File: builds/linux/debian/11/linux-debian.pkr.hcl [hunk 2]
```diff
@@ -139,7 +139,8 @@ build {
     playbook_file = "${path.cwd}/ansible/main.yml"
     roles_path    = "${path.cwd}/ansible/roles"
     ansible_env_vars = [
-      "ANSIBLE_CONFIG=${path.cwd}/ansible/ansible.cfg"
+      "ANSIBLE_CONFIG=${path.cwd}/ansible/ansible.cfg",
+      "ANSIBLE_PYTHON_INTERPRETER=/usr/bin/python3"
```
[the remark this answers is not in the document]
Comment (tenthirtyam) [reply builds/linux/debian/11/linux-debian.pkr.hcl:144]: It suppresses the auto-detect warning.

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
The `locals` for the `data_source_command` will need to be added and tested.

Example:

```hcl
locals {
  #....
  data_source_command = var.common_data_source == "http" ? "url=http://{{ .HTTPIP }}:{{ .HTTPPort }}/ks.cfg" : "file=cdrom:/ks.cfg"
}
```

**Comment by tenthirtyam** [comment]
Hi @evilaliv3, could you resolve the conflicts that have arisen after the v22.05 release?

I'll target this one for v22.06 once I get the Ansible tasks updated, which seem fairly minor.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

**Comment by stefangweichinger** [comment]
Great work, testing myself here.

**Comment by tenthirtyam** [comment]
> **Note**
> 
> Guest OS identifier will need to be updated due to [KB 85845](https://kb.vmware.com/s/article/85845).

**Comment by tenthirtyam** [comment]
Guest resets to Debian 11 during clone and fails customization.

**Comment by tenthirtyam** [comment]
For now, moving from `v22.07` milestone to the `Backlog` as the cloning issue needs to be resolved for Debian 11-based OVF templates.

**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

**Comment by evilaliv3** [comment]
Thank you @tenthirtyam for the update.

Which are the current issues that you would like to fix?

**Comment by tenthirtyam** [comment]
Hey @evilaliv3 - the outstanding items include:

- the Ansible playbook updates - simple, will address that.
- update the `disk`-based data source option for kickstart to use the methods mentions earlier in the thread, but retain the `http`-based option by default.
- there's an open internal engineering issue for VM clone of Debian 11 for guest customization. OS needs to be set to other4xLinux64Guest by default per KB, but when cloned from content library it automatically changes it to Debian 11 and customizations fail. I filed and am tracking a fix internal for a vCenter Server update.


**Ryan Johnson**
Staff II Solutions Architect | VMware, Inc.

**Comment by tenthirtyam** [comment]
1. Regarding:

> _*...there's an open internal engineering issue for VM clone of Debian 11 for guest customization. OS needs to be set to other4xLinux64Guest by default per KB, but when cloned from content library it automatically changes it to Debian 11 and customizations fail. I filed and am tracking a fix internal for a vCenter Server update."_

I have a fix for this issue that will be addressed in the Ansible playbook.

TL;DR: 

`/etc/open-vm-tools/tools.conf`

```vim
[guestosinfo]
short-name = other4xLinux64Guest
```

2. Regarding:

>_"...update the disk-based data source option for kickstart to use the methods mentions earlier in the thread, but retain the http-based option by default."_

This will be HTTP-only for the first inclusion, much like SLES 15. We can add the `disk` method later.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Comment by tenthirtyam** [comment]
Merged! 🚀 

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.
